### PR TITLE
new version

### PR DESCRIPTION
**Pull request by cainus**
* separates getters / setters for stronger typing.  eg. `port()` is used for setting the port, but `getPort()` is used for getting it.  UrlGrey is more for url-manipulation than parsing, so `setPort()` seemed too verbose and backward incompatible.
* uses typescript & jest
* drops explicit browser support
* uses native WHATWG classes and drops node.js imports and 3rd-party deps .  These should be faster.

### CODE REVIEW

**Comment by KelWill** [comment]
I'm not seeing the new TS files. Did you skip adding them to the commit?

**Comment by cainus** [comment]
> 

Bah yeah sorry... they're on the branch now.

File: index.ts [hunk 1]
```diff
@@ -0,0 +1,486 @@
...
+
+const getDefaults = function () {
+  const defaultUrl = "http://localhost:80"
+  const defaults = urlParse(defaultUrl)
+  return defaults
+}
+
+export class UrlGrey {
+  #url: string
+  _parsed: Parsed | null
```
Comment (bttmly) [inline index.ts:62]: is this intended to be public? it has an underscore but no visibility modifier

File: index.ts [hunk 2]
```diff
@@ -0,0 +1,486 @@
...
+
+  parsed() {
+    if (this._parsed) {
+      return this._parsed
+    }
+    this._parsed = urlParse(this.#url)
+    const defaults = getDefaults()
+    
+    this._parsed.protocol = this._parsed.protocol || defaults.protocol
+    return this._parsed
```
Comment (bttmly) [inline index.ts:77]: We should mark this as `readonly` in typescript, or copy it on the way out. readonly is probably better. if we don't do one of these then calling code can change the internal state by mutating the return value.

File: index.ts [hunk 3]
```diff
@@ -0,0 +1,486 @@
...
+    clone.parsed().query = p.query
+    return clone.path(p.pathname)
+  }
+
+  port(num: number | string | null) {
+    if (this.getProtocol() === "file") {
+      throw new Error("file urls don't have ports")
+    }
+    const clone = this.clone()
+    clone.parsed().port = parseInt(`${num}`, 10)
```
Comment (bttmly) [inline index.ts:100]: ah I see you're relying on manipulating the internal state here. hmm. Maybe we should instead have a method called `extend` which takes a `Partial<Parsed>` merges it with the current instance's `Parsed` and produces a new `UrlGrey` from that.

File: index.ts [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+  username: string;
+  password: string;
+  host: string;
+  hostname: string;
+  hash: string | null;
+  search: string;
+  pathname: string;
+  port: number | null;  // some protocols have no port, like file://
+  path: string;
+  query: string;
```
Comment (bttmly) [inline index.ts:20]: any reason not to use URLSearchParams here for the `query` property?

File: index.ts [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+    href += this.getHash() ? "#" + this.getHash() : ""
+    return href
+  }
+
+  extendedPath(url: string) {
+    const clone = this.clone()
+    const p = urlParse(`http://localhost/${url}`)
+    clone.parsed().hash = p.hash
+    clone.parsed().query = p.query
+    return clone.path(p.pathname)
```
Comment (bttmly) [inline index.ts:92]: is there any rhyme or reason to which fields require you to go through `this.parsed().xyz` vs having a dedicated method like `this.xyz()`?

File: package.json [hunk 6]
```diff
@@ -5,60 +5,34 @@
...
-  "browser": {
-    "url": "./browser/url.js",
-    "querystring": "./browser/querystring.js"
+    "test": "jest"
   },
   "maintainers": [
     "Gregg Caines <gregg@caines.ca> (http://caines.ca)"
   ],
-  "dependencies": {
-    "fast-url-parser": "^1.1.3"
```
Comment (KelWill) [inline package.json:23]: I'd recommend benchmarking this removal. There's a huge difference between it and the Node.JS url library. Using fast-url-parser significantly reduced how much event-loop-blocking we were seeing within API when handling 1,000-10,000s of urls.

**Review by KelWill — COMMENTED** [review]
`URL` feels like it already supports many of the features that urlgrey is adding on.

- would it make sense to have `_parsed` be a URL object that you can mutate and stringify?
- does the library as a whole still make sense? `URL` has good documentation and is pretty easy to work with! If there are functions that URL is missing, would it make sense to set up helpers that take a `URL` object and transform that object?

**Pull request merged**